src/NHibernate/Type/GenericOrderedSetType.cs

```diff
@@ -24,9 +27,41 @@ public GenericOrderedSetType(string role, string propertyRef)
 		{
 		}
 
+		static ConstructorInfo SetConstructor
+		{
+			get
+			{
+				if (_setConstructor == null)
+				{
+					lock (typeof(GenericOrderedSetType<>))
```

Comment:
Lock should be put on privately held fields for avoiding deadlock in case some code elsewhere put a lock on the same public thing. (Msdn recommended practice, with some discussion about it [here](http://stackoverflow.com/q/1603008/1178314).) Illustrated in this [PR](https://github.com/nhibernate/nhibernate-core/pull/590/files) by the way.